src/symbol/quads.js

```diff
@@ -265,62 +302,123 @@ export function getGlyphQuads(anchor: Anchor,
                 // By rotating counter-clockwise around the point at the center of the left
-                // edge of a 24x24 layout box centered below the midline, we align the center
-                // of the glyphs with the horizontal midline, so the yOffset is no longer
+                // edge of a 24x24 layout box centered below the midline, we align the midline
+                // of the rotated glyphs with the horizontal midline, so the yShift is no longer
                 // necessary, but we also pull the glyph to the left along the x axis.
-                // The y coordinate includes baseline yOffset, thus needs to be accounted
-                // for when glyph is rotated and translated.
-                const center = new Point(-halfAdvance, halfAdvance - SHAPING_DEFAULT_OFFSET);
+                const yShift = (positionedGlyph.y - currentOffset);
+                const center = new Point(-halfAdvance, halfAdvance - yShift);
                 const verticalRotation = -Math.PI / 2;
-
-                // xHalfWidthOffsetCorrection is a difference between full-width and half-width
-                // advance, should be 0 for full-width glyphs and will pull up half-width glyphs.
-                const xHalfWidthOffsetCorrection = ONE_EM / 2 - halfAdvance;
-                const yImageOffsetCorrection = positionedGlyph.imageName ? xHalfWidthOffsetCorrection : 0.0;
-                const halfWidthOffsetCorrection = new Point(5 - SHAPING_DEFAULT_OFFSET - xHalfWidthOffsetCorrection, -yImageOffsetCorrection);
                 const verticalOffsetCorrection = new Point(...verticalizedLabelOffset);
-                tl._rotateAround(verticalRotation, center)._add(halfWidthOffsetCorrection)._add(verticalOffsetCorrection);
-                tr._rotateAround(verticalRotation, center)._add(halfWidthOffsetCorrection)._add(verticalOffsetCorrection);
-                bl._rotateAround(verticalRotation, center)._add(halfWidthOffsetCorrection)._add(verticalOffsetCorrection);
-                br._rotateAround(verticalRotation, center)._add(halfWidthOffsetCorrection)._add(verticalOffsetCorrection);
+                // Relative position before rotation
+                // tl ----- tr
+                //   |     |
+                //   |     |
+                // bl ----- br
+                tl = new Point(-halfAdvance + builtInOffset[0], builtInOffset[1]);
+                tl._rotateAround(verticalRotation, center)._add(verticalOffsetCorrection);
+
+                // Relative position after rotating
+                // tr ----- br
+                //   |     |
+                //   |     |
+                // tl ----- bl
+                // After rotation, glyph lies on the horizontal midline.
+                // Shift back to tl's original x coordinate before rotation by applying 'xOffsetCorrection'.
+                const offsetCorrection = new Point(-yShift + halfAdvance, 0);
+                tl._add(offsetCorrection);
```

Comment:
Nit: simplify to avoid `Point` allocation:

```suggestion
                tl.x += -yShift + halfAdvance;
```